crates/wash-lib/src/run/ctx.rs

```diff
@@ -0,0 +1,113 @@
+use crate::registry::OciPullOptions;
+use std::path::PathBuf;
+use wasmtime_wasi::{WasiCtx, WasiCtxBuilder};
+
+/// A structure bearing information about:
+/// - Where and how to find the binary of the component to run
+/// - Which parts of the host the component may access
+pub struct Ctx {
+    pub(crate) oci_pull_options: OciPullOptions,
+    pub(crate) oci_cache_file: Option<PathBuf>,
+    pub(crate) wasi_ctx: WasiCtx,
+    pub(crate) reference: String,
+    // TODO(raskyld): Add `wasi:http` support once basic WASI WIT is supported.
+}
+
+/// Builder-style structure used to create a [`Ctx`].
+pub struct CtxBuilder {
+    wasi: WasiCtxBuilder,
+    built: bool,
+    oci_pull_options: Option<OciPullOptions>,
+    oci_cache_file: Option<PathBuf>,
+    reference: Option<String>,
+}
+
+impl CtxBuilder {
+    /// Creates a builder allowing to programmatically define access policies.
+    ///
+    /// Once you defined your access policies, use the [`Self::build`] function to acquire a [`Ctx`].
+    ///
+    /// The [`Ctx`] can, then be passed to the [`LocalRuntime::run`][runtime-run] function to
+    /// run a component while enforcing those access policies.
+    ///
+    /// [runtime-run]: crate::run::LocalRuntime::run
+    pub fn new() -> Self {
+        CtxBuilder {
+            wasi: WasiCtxBuilder::new(),
+            built: false,
+            oci_pull_options: None,
+            oci_cache_file: None,
+            reference: None,
+        }
+    }
+
+    /// Builds the final [`Ctx`].
+    ///
+    /// This method can be called only once. Calling it again will panic.
+    ///
+    /// # Panics
+    ///
+    /// Panics if called more than one time.
+    pub fn build(&mut self) -> anyhow::Result<Ctx> {
+        assert!(!self.built);
+
+        self.built = true;
+
+        if self.reference.is_none() {
+            return Err(anyhow::anyhow!("reference cannot be empty"));
+        }
+
+        anyhow::Ok(Ctx {
+            wasi_ctx: self.wasi.build(),
+            oci_pull_options: self.oci_pull_options.take().unwrap_or_default(),
+            oci_cache_file: self.oci_cache_file.take(),
+            reference: self.reference.take().unwrap(),
+        })
+    }
+
+    /// Returns a [`WasiCtxBuilder`] that can be used to control access to most WASI
+    /// native interfaces.
+    pub fn wasi_ctx(&mut self) -> &mut WasiCtxBuilder {
```

Comment:
💡 I am not sure whether we should expose the `WasiCtxBuilder` like that or decorates it with our own functions instead.

This approach is easy and simple but it forces me, for example, to re-publish parts of the wasmtime crate like the `wasmtime_wasi::{DirPerms, FilePerms}` to avoid `wash-cli` a direct dependency on wasmtime.

Reply:
Yeah wasmtime updates enough I'm not sure I'd want to expose this as part of our public API. However, I also don't mind bringing the wasmtime dep to wash-cli since we already bring it in transitively via wash-lib. Let me do a full review and see if I have any other ideas